src/runtime/storage/storage-provider-factory.ts

```diff
@@ -69,13 +72,25 @@ export class StorageProviderFactory {
     return await this._storageForKey(key).connect(id, type, key);
   }
 
+  private async _acquireMutexForKey(key: string) {
+    if (!this.mutexMap.has(key)) {
+      this.mutexMap.set(key, new Mutex());
+    }
+    return this.mutexMap.get(key).acquire();
+  }
+
   async connectOrConstruct(id: string, type: Type, key: string) : Promise<StorageProviderBase> {
     const storage = this._storageForKey(key);
-    let result = await storage.connect(id, type, key);
-    if (result == null) {
-      result = await storage.construct(id, type, key);
+    const release = await this._acquireMutexForKey(key);
```

Comment:
this method is not longer used in the StorageStub::inflate method, so this change is irrelevant for this PR, and not practically needed. However it does seem the right thing to do - keep it or drop it?

Reply:
what does this protect against? Is it re-entrance during speculative execution?

Reply:
yes, but i was only hitting problems here when i was using the `connectOrConstruct` method from the volatile storage shim inflate method. and i don't anymore. so technically, this is not needed now. but theoretically i think, it is the right thing to do.

Reply:
yeah, I think connect is re-entrant, but construct isn't.. :)

I'm OK with this staying here.